flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCheckpointPlanCalculator.java

```diff
@@ -111,7 +112,10 @@ public void setAllowCheckpointsAfterTasksFinished(boolean allowCheckpointsAfterT
                                         ? calculateAfterTasksFinished()
                                         : calculateWithAllTasksRunning();
 
-                        checkTasksStarted(result.getTasksToTrigger());
+                        checkTasksStarted(
+                                isUnalignedCheckpoint
+                                        ? result.getTasksToWaitFor()
+                                        : result.getTasksToTrigger());
```

Comment:
Do I understand correctly, that the motivation to limit the change to UC mode only is that there is channel state to recover in UC mode?

If so, I think it would be better to always use `result.getTasksToWaitFor()`, because
1. operator state initialization is a part of restore/init phase too
1. UC mode option can change but the snapshot can still contain channel state that needs to be restored
WDYT?

Reply:
Actually, I agree with you. But as I understand, @AHeise has different opinion?

Reply:
The main motivation is the following:
- imagine you are running a large-scale application in AC
- you checkpoint every minute because of SLA
- after some failure, Flink is in recovery but recovery takes a 2 minutes
- we have implicit backpressure as any downstream task with large state will block upstream

previously:
- checkpoint barriers have been added to the stream while recovering
- as soon as a task finishes recovery, it can also start checkpointing

with the change (and no distinction):
- during that time no checkpoints are triggered
- only after every task has been recovered, is checkpoint started at all
- the first checkpoint is delayed compared to before

However, after having written that, I'm not too sure if the new behavior is that bad:
- the checkpoints that is generated previously is awfully close to the recovered checkpoint unless there is huge backpressure
- having multiple checkpoint barriers being backpressured during recovery might lead to quite a bit of I/O to the checkpoint storage after the last task has recovered.

So all in all, only high backpressure cases would really suffer from that change and we want to encourage UC for them. High throughput, large scale with AC might actually benefit from that changed behavior.

Reply:
@AHeise, So as I understand from your comment, in general, you don't have any objections and we can use `result.getTasksToWaitFor()` for both cases(AC and UC)?

Reply:
What @AHeise described makes sense to me (skipping processing of data accumulated during recovery and passing barriers right away, reducing e2e latency) but it seems quite a narrow use-case:
- checkpoint timeout must be less than recovery time
- time to process the accumulated data must be significantly higher than checkpoint interval

So one alternative would be to add an option to allow power-users to trigger checkpoints while still in recovery, but don't trigger by default. But such option can also be added separately.
WDYT?

Reply:
@rkhachatryan 
> checkpoint timeout must be less than recovery time

Do you mean 'checkpoint timeout must be **greater** than recovery time'? If I understand checkpoint timeout correctly, then the checkpoint will be discarded and it will start from beginning if checkpoint timeout is less than recovery time. 

> one alternative would be to add an option to allow power-users to trigger checkpoints while still in recovery, but don't trigger by default.

I don't think that a having an extra option for power-user is a good choice. Because as you mentioned by yourself it is narrow use-case and actually I don't see the significant advantages in this aproach. So supporting of new feature for pretty limited scenarios and with no obvious explanation doesn't seem good.


I see only a couple of advantages of starting the checkpoint ASAP(Arvid has mentioned about it already):

- Sometimes(depends on bottle neck in recovery) the first checkpoint can finish earlier.
- The first checkpoint can do its work before job really started which can influent processing of data in less degree.

In my opinion, the only first point makes sense. It is better to have the first checkpoint sooner rather than later. But I still don't understand how it is important because for UC which we want to have as primal one, we don't support such behaviour. So my position is to have the same behaviour for both checkpoints. If we think that the delay in starting of the first checkpoint is crucial then we should support it for UC(maybe not in this ticket but in general). but if we think that it is not so important then we can remove this support from AC.

So in general, it seems to have this feature can be helpful but perhaps, we should think about having the same feature for UC.

Reply:
> So all in all, only high backpressure cases would really suffer from that change and we want to encourage UC for them.

> In my opinion, the only first point makes sense. It is better to have the first checkpoint sooner rather than later. But I still don't understand how it is important because for UC which we want to have as primal one, we don't support such behaviour. So my position is to have the same behaviour for both checkpoints. If we think that the delay in starting of the first checkpoint is crucial then we should support it for UC(maybe not in this ticket but in general). but if we think that it is not so important then we can remove this support from AC.

I think those are good arguments.

However there is also one more case which I would be worried about. What if recovery is a thing that's causing this huge backpressure, and apart of that, user is happy by using AC? For example recovery takes 1h, checkpointing interval is 10minutes, recovery creates back pressure worth of 1h (every minute sources are running while some downstream task is not processing records is creating a minute backlog of records). Currently first checkpoint would complete after 1h 10 minutes. If we wait for recovery to complete before triggering first checkpoint, first checkpoint would complete after 2h.

Reply:
> In my opinion, the only first point makes sense. It is better to have the first checkpoint sooner rather than later. But I still don't understand how it is important because for UC which we want to have as primal one, we don't support such behaviour. So my position is to have the same behaviour for both checkpoints. If we think that the delay in starting of the first checkpoint is crucial then we should support it for UC(maybe not in this ticket but in general). but if we think that it is not so important then we can remove this support from AC.

UC will not have this issue as strongly as the barrier travels pretty much instantly as soon as recovery finishes. The slowest part during UC (biggest state) will probably also be the part that takes longest to recover.

The worst part is that since UC overtakes, we will cancel all checkpoints until the job fully recovered. This can put a huge load onto the checkpoint storage and I/O in general. Hence, we absolutely need this change for UC.

> However there is also one more case which I would be worried about. What if recovery is a thing that's causing this huge backpressure

The big question here is: wouldn't we prolong recovery and increase backpressure if we write data to checkpoint storage at the same time as we recover from it if I/O is the bottleneck? I'm assuming there is no clear answer to it. I'm also reluctant to shift the responsibility to the user because I don't know how any non-expert can estimate that.

That is, I'm assuming both ways are fine with their disadvantages and we should just pick one. In doubt, pick the one that preserves the old behavior.

Reply:
@akalash 
> Do you mean 'checkpoint timeout must be greater than recovery time'? If I understand checkpoint timeout correctly, then the checkpoint will be discarded and it will start from beginning if checkpoint timeout is less than recovery time.

@pnowojski 
>  Currently first checkpoint would complete after 1h 10 minutes. If we wait for recovery to complete before triggering first checkpoint, first checkpoint would complete after 2h.

Given that the checkpoint timeout is big enough (default is 10m, for larger setups I saw 30m). 
(I think it also answers @akalash question - yes, it's about wording). 

Yes (it's a matter of wording). 
Timeouts are counted as checkpoint failures, and if this count reaches a threshold then the job will fail.

@AHeise 
> That is, I'm assuming both ways are fine with their disadvantages and we should just pick one. In doubt, pick the one that preserves the old behavior.

In this case old behavior (for AC) is combined with the new behaviour (for UC) and this seems misleading from both users and developers point of view IMO.

Reply:
> In this case old behavior (for AC) is combined with the new behaviour (for UC) and this seems misleading from both users and developers point of view IMO.

Also a fair point. I don't have a strong opinion on this anymore as I think both versions are good (and a bit bad). If you say that, it's easier for the user (not 100% clear to me) and dev to unify it, then I'm for it.

Reply:
ok, +1 from my side on unifying the behaviour between UC and AC.

One more question from my side. Is this change compatible with FLIP-147? When some of the sources have already or are about to `FINISHED`? `calculateAfterTasksFinished` a couple of lines above suggest so, but can you double check?

Reply:
Good point, I also think that checking `taskRunningStatus` in `calculateAfterTasksFinished` should be enough.

@gaoyunhaii, could you please confirm that replacing `checkTasksStarted(result.getTasksToTrigger())` with `checkTasksStarted(result.getTasksToWaitFor());` is fine here?

Reply:
Hi @rkhachatryan, it should be ok since the finished tasks would not be included in `result.getTasksToWaitFor()`, only running tasks would.